resholve: 0.9.1 -> 0.10.2, ongoing cross fixes

abathur opened the pull request:
## Description of changes

Continuing from #297626 due to retarget pings (see approvals there; cc @danielbarter @K900 @thiagokokada).

AFAICT this will fix #295072 and fix #292393. I've tested cross for `lesspipe` and `xdg-utils` in CI (https://github.com/abathur/resholve/actions/runs/8474467054/job/23238072055).

I'm not certain about the approach here; if this is wrong I could use a little help understanding why. I've left review comments to annotate most of this to make it easier to see what's actually cross-related.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

abathur committed:
resholve: 0.9.1 -> 0.10.2, ongoing cross fixes

thiagokokada commented on pkgs/development/misc/resholve/oildev.nix:
```diff
@@ -58,59 +38,54 @@ rec {
       */
       postFetch = ''
-        rm -rf $out/{Python-2.7.13,metrics,py-yajl,rfc,gold,web,testdata,services,demo,devtools}
+        rm -rf $out/{Python-2.7.13,metrics,py-yajl,rfc,gold,web,testdata,services,demo}
```
Small nit here, but could we extract the Python version from the package itself?

Alternatively, we could use `Python-*`.

abathur replied:
This is just oil's vendored python

abathur commented:
Going back to draft to evaluate the error reported in https://github.com/NixOS/nixpkgs/issues/292393#issuecomment-2021877222.

abathur committed:
xdg-utils: override lore for perl

The OSH update (and associated refactoring) in resholve seem
to have uncovered surfaced a fresh error here.

abathur commented:
@danielbarter Can you give this another try?

danielbarter commented:
yep, after rebasing this into staging, both `lesspipe` and `xdg-utils` build.

abathur commented:
I left annotations on the previous PR (https://github.com/NixOS/nixpkgs/pull/297626#pullrequestreview-1956172613). Only material differences here should be the updated resholve rev and the lore tweak for xdg-utils.

danielbarter reviewed (approved):
Can't really comment on the resholve related details, but can confirm that both `xdg-utils` and `lesspipe` cross compiling (at least for aarch64) are fixed by this.

More generally, i think the resholve approach to improving the reliability of shell script packaging makes sense, but I am a little suspicious about the oil dependency. It is a fairly large project, with a vendored python 2 dependency (which is compiled to C++?!). If this is going to become the standard way for patching shell scripts in nixpkgs, it would be good to know that oil has a long term plan towards becoming, more boring?

abathur commented:
> it would be good to know that oil has a long term plan towards becoming, more boring?

<details><summary>I guess this depends on your definition of boring :) (hiding since this is a tangent)</summary>

From resholve's prespective, I want a parser that is of sufficiently-high quality to drive a working shell, and ideally one that tries to do all it can statically at parse time. 

From this perspective, a very-complete parser that is good about failing at parse time when it's confused is quite boring compared to a much-less-complete parser that enables us to silently and subtly corrupt scripts as we rewrite them. 

I am not sure I would have arrived at them myself, but i think Andy's made a persuasive defense of his technical choices (and the ground he's covered also attests). 

As of the latest (0.21.0) release the c++ translation is finally passing all of the spec tests that the python impl passes. We still have to solve for interchange, whether thats just passing a parse tree representation or writing a python extension to drive the parser--but in either case it should enable resholve itself to get more boring by moving off of the python2 implementation.
</details>
